### DISCUSSION TITLE
Feature Wishlist

### DISCUSSION THREAD

**Discussion by hannobraun** [hannobraun/fornjot#146] [Ideas]
## The feature wishlist is deprecated

This wishlist was created back when I was working on Fornjot full-time. I am no longer able to do that, and development has slowed down (compared to the early years) as a result.

To better reflect that reality, I'm deprecating this feature wishlist. Some of the things in here might still make it into Fornjot, but most of it should be considered out of scope for the time being, and only foundational features are considered in scope.

If you are interested in working on any of the items listed here, please start doing so in a third-party crate. Any changes to Fornjot's API that are required for you to do so are considered in scope. But any non-foundational features should be proven out in third-party crates, before they can be considered for merging.

Aside from adding this notice, I've left the rest of the list as it was.

---

This is the feature wishlist for Fornjot! If there's anything you'd like to see in Fornjot, please comment here. I'll keep this post up-to-date with suggestions posted below.

## Actionable

We keep track of feature requests that are actionable, i.e. can be worked on in the near-term, using [**GitHub Issues**](https://github.com/hannobraun/Fornjot/issues). If you're not sure whether your feature request falls into that category, feel free to just post here and ask.

## Would be cool, but...

These are feature requests that would be cool to have, aren't really blocked by anything (so could be worked on in the near term), but there's another problem. Like unclear demand (someone mentioned it or it seems obvious, but unclear if anyone would actually use it), or lack of resources (nobody has expressed interest in working on it).

If you're interested in any of these features, please speak up! It doesn't take a lot to move them to the "actionable" list.

- **SVG/DXF/... export:** Export sketches to 2D formats like SVG or DXF. This is something that could be worked on, but with big changes to the geometry representation incoming (https://github.com/hannobraun/fornjot/issues/2118), that exporting functionality would have to be rewritten, essentially. Also see https://github.com/hannobraun/fornjot/issues/798.
- **SVG/DXF/... import:** Import an SVG, DXF, and other 2D formats to create a sketch. Could be considered premature, as sketches are too limited to properly support those formats, but a preliminary version that that converts curves into line segments could be done right now.
- **3MF/STL/... import:** Import 3MF, STL, or other mesh-based formats into Fornjot.
- **More mesh-based export formats:** Export to glTF and other file formats. See https://github.com/hannobraun/Fornjot/issues/155.
- **Textures:** For some use cases, it would be useful to add textures to faces. So far, Fornjot has been too basic in its capabilities to really needs something like that, but if someone was interested, it's something that can be worked on. (As a middle ground between the current state, a face having one color, and textures, it could make sense to have other options, like defining gradients. I'm not going to add a separate item for that, as it's tightly related to this one.)
- **Sketching on faces:** Sketching on an arbitrary face, then extruding that sketch or creating a pocket based on it, is an essential workflow. There are a few building blocks required for that, like splitting the face in more complex ways than is currently possible, and making the sweeping infrastructure more flexible. If anyone wanted to start a concerted effort to tackle this right now, it would be possible. If not, it makes more sense to wait until more of the building blocks are in place.
- **Measuring generated geometry:** It would be great to be able to take measurements off generated geometry. It's currently unclear how exactly this would look like, so some design work is required. But it doesn't sound like something that is insurmountable.
- **Graphical validation errors:** Validation errors are currently displayed textually, which can make it quite hard to figure out what they refer to. It would be much better to display them graphically, as part as the normal view of the model. While this would be highly useful, it is also a big project. It would need someone committed to really drive it forward.

## Premature

Premature feature requests are those that we'd like to have, but that aren't ready to be worked on, for one reason or another.

Some of those might be pretty close to ready though. A lot of them require research to figure out what's even required to get them done, so they might not even be that hard to do. If you're interested in a specific items, feel free to do that research. It might be all that's required to move a feature to the "actionable" list.

- **Technical drawings:** Exporting technical drawings should be possible. It's probably a bit too early for that, while the CAD kernel is still so much in flux. See https://github.com/hannobraun/Fornjot/issues/2.
- **Chamfers/fillets:** This is, by all appearances, a non-trivial feature (at least to get working reliably). Right now the system for representing geometry is quite primitive, and I think we should wait for it to mature before tackling chamfers/fillets.
- **Bézier curves or b-splines or NURBS for sketches**: The CAD kernel is still under heavy development and things are changing all the time. For now we should focus on a [simpler feature set](https://github.com/hannobraun/Fornjot/milestone/1) and reconsider once things have settled down somewhat.
- **Bézier curves or b-splines or NURBS for surfaces**: See previous entry. 
- **Animation:** Many models include moving parts. Being able to visualize that, with some basic assistance like collision-detection, should be possible. This is another one of those features that require the CAD kernel to settle down a bit first, before it can be considered.
- **Export to STEP:** I'm not sure what would be involved in that, but I suspect it's another of those features that require the CAD kernel to settle down a bit.
- **PCB import:** When designing products that involve electronics, it would be useful to import PCBs, to get their dimensions, hole locations, etc.
- **Architecture:** Support architecture use cases, including stuff like [IFC](https://en.wikipedia.org/wiki/Industry_Foundation_Classes). I have no expertise here, and so far it's largely unclear what specifically would be required.
- **Assemblies:** Create assemblies that are made up of multiple components. Display assemblies as exploded views. See #220 for more context.
- **Assembly instructions:** Generate assembly instructions for your Fornjot model. This could be supported by making it easy to embed Fornjot models in websites. Not just as a full 3D view, but by making it easy to generate pictures of specific parts of the model, from specific perspectives, for example. Maybe it makes sense to teach Fornjot itself about assembly instructions on top of that, but I haven't thought that through. See #220 for more contex.
- **Cut lists:** Teach Fornjot about materials and the sizes they come in, so it can generate cut lists for you. See #220 for more contex.
- **STEP/DWG/... import:** The kernel isn't ready to support these formats yet.
- **More flexible sweeps:** It's already possible to sweep some objects along a straight path. It should be possible to sweep along a circle, helix, or even arbitrary (defined by an arbitrary function) paths. It probably makes no sense to attempt this, before we have a more powerful geometry representation (https://github.com/hannobraun/fornjot/issues/2118). Also see https://github.com/hannobraun/Fornjot/issues/997.
- **Constraint-based sketches:** This would be very hard to implement with the current data structures. I think we need more flexible ways to represent geometry before it makes sense to tackle this.
- **2D and 3D boolean operations:** This is something I had already been working on (see https://github.com/hannobraun/fornjot/issues/42, https://github.com/hannobraun/fornjot/issues/43, https://github.com/hannobraun/fornjot/issues/44; also https://github.com/hannobraun/fornjot/issues/943), but there are *a lot* of moving parts to it. Both on the "shape manipulation" side, but also on the "information gathering" side (lots of intersection tests). I came to the conclusion that it makes more sense to build up geometry manipulation (which provides useful benefits in itself), before tackling this feature again.

## Out of scope

For a project to work, it needs to have some kind of vision to guide it. Some features don't fit that vision, and are thus considered out of scope. What's considered out of scope can change, as the vision and the resources to realize it evolve.

Maybe some of those could start as external projects and later be merged into the project. If you're interested in starting such an effort (whether the goal is to eventually merge it or not), please speak up. Making Fornjot more flexible, so an ecosystem can be built around it, is an important goal, and making changes that enable such work is always on the table.

- **CAM/slicer features**: I'd love to one day support generating G-Code directly, by integrating CAM features for machining and slicer features for 3D printing. This would of course use a programmatic approach, like [FullControl GCODE Designer](https://fullcontrolgcode.com/), but integrated with Fornjot's CAD features.
  Creating a CAD kernel is already enough work, however. We'd have to make real headway in that area before something like this could be considered.
- **Simulation:** Full-scale simulation to test models would be awesome, but again, CAD is already enough work.
- **PCB design:** Adding support for designing PCBs is something I have though about, but so far it's totally unclear to me what a code-based approach could add here. But it could be interesting to design PCBs and the products that feature them using the same tool.

**Comment by hannobraun** [comment]
Added "Configuration" as premature feature.

**Reply by hannobraun** [reply]
Added "PCB import", "PCB design", and "Sketching on surfaces".

**Reply by hannobraun** [reply]
Added "WASM add-ons".

**Reply by hannobraun** [reply]
Added "Would be cool, but..." category and added a few items. Added "Architecture" to "premature" list.

**Reply by hannobraun** [reply]
Added *Assemblies*, *Assembly instructions*, and *Cut lists*.

**Reply by hannobraun** [reply]
Added *Option to switch off automatic model reloading* and *Load models from within Fornjot*.

**Reply by hannobraun** [reply]
Added *Load model from current directory*.

**Reply by hannobraun** [reply]
Added *More mesh-based export formats* and *Building generators*.

**Reply by hannobraun** [reply]
Added *Packaging*.

**Reply by hannobraun** [reply]
Added *More fine-grained ecoystem*.

**Reply by hannobraun** [reply]
Added *Display far-off background*.

**Reply by hannobraun** [reply]
Added *Build models from b-rep building blocks*.

**Reply by hannobraun** [reply]
- Changed *SVG import* into *SVG/DXF/... import*.
- Added *STEP/DWG/... import*.
- Added *2D transformations*.
- Added *Measuring generated geometry*.

**Reply by hannobraun** [reply]
Removed *2D transformations*, which is now an issue (#944).

**Reply by hannobraun** [reply]
- Added *Helical sweeps*
- Added *Sweeps along arbitrary paths*

**Reply by hannobraun** [reply]
Moved *Constraint-based sketches* to *Would be cool, but...*

**Reply by naggie** [reply]
Glad to see STEP export on the list!  IMO it's the Achilles heel of OpenSCAD, making it difficult to manufacture on anything other than a 3D printer. Most CAM software for subtractive machining requires a STEP file.

I guess the challenge, like DXF, will be to ensure things like arcs are used correctly instead of an approximation; though perhaps this is a non-issue with the Fornjot kernel unlike openscad, where everything is a triangle based approximation on render.

**Reply by hannobraun** [reply]
Thanks for the feedback, @naggie!

> I guess the challenge, like DXF, will be to ensure things like arcs are used correctly instead of an approximation; though perhaps this is a non-issue with the Fornjot kernel unlike openscad, where everything is a triangle based approximation on render.

I believe it is a non-issue from an architectural perspective. The Fornjot kernel uses b-rep data structures internally, and triangle meshes are just created as an output, for rendering and 3MF/STL export. So STEP export would, theoretically, just be a matter of serializing the internal data structures to that specific format.

It's still a practical problem though. I haven't looked into the details of STEP yet, so it might not be that easy to support in practice. Plus, the Fornjot kernel is still so limited, from a feature perspective, that you can't do a whole lot, STEP or not. But that's just a matter of putting in the work.

**Reply by naggie** [reply]
Thanks for the lightning quick reply! Glad to hear the architecture would allow for it.

By the way I'm super impressed with the momentum of Fornjot. I have no idea how you have the patience to develop so fast! Thanks for the hard work

**Reply by hannobraun** [reply]
Thank you! Progress certainly doesn't feel fast to me, even though a lot is happening week by week :smile: 

I'm in the fortunate situation that I can focus on this project full-time, which explains both the hard work and the lightning-quick replies (at least during the day, I spend a lot of time on GitHub). That's all thanks to my sponsors, of course, so *thank you* for being a part of that!

**Reply by naggie** [reply]
Oh that's a nice situation to be in. I'd like to do the same one day with my projects. Happy to sponsor -- I figured it was better than just starring even if it is the lowest amount

**Reply by hannobraun** [reply]
Added *Textures*.

**Reply by hannobraun** [reply]
This wishlist had really fallen out of use since [the project switched focus](https://www.fornjot.app/blog/a-new-direction/). So I did a big update, removing some stuff, and moving other stuff around. Surprisingly (at least to me) I was able to move some stuff from *Premature* to *Would be cool, but...*. Seems like the project *is* making progress, even though it doesn't always feel like that to me, being down in the trenches :smile:

**Reply by hannobraun** [reply]
Added *Boolean operations* to *Would be cool, but...*.

**Reply by hannobraun** [reply]
- Added *SVG/DXF/... export* to *Would be cool, but...*
- Expanded entry on boolean operations to include both 2D and 3D
- Moved *2D and 3D boolean operations* to *Premature*
- Merged entries about sweeping into a single one, *More flexible sweeps*; also added circular sweeps there

**Reply by hannobraun** [reply]
Added *Graphical validation errors*.

**Comment by hannobraun** [comment]
This feature wishlist is now deprecated. See the note I added to the top of the original post.
